controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+    // those related to waiting for new messages on `outgoing_request_rx`. The
+    // effect is that we only wait for new outgoing responses if we're not
+    // currently processing one.
+    let mut outstanding_request: Option<OutstandingHostRequest> = None;
+
+    loop {
+        // We also need to resend an outstanding request, as long as there is
+        // one. This future is recreated on each pass through the loop.
+        let retry_timeout = if let Some(req) = &outstanding_request {
+            sleep_until(req.last_sent + RESEND_INTERVAL)
+        } else {
+            // NOTE: This will never be polled, but the future is evaluated in
+            // each branch of `tokio::select!`, even if it's eventually disabled
+            // by the condition on it.
+            sleep_until(Instant::now() + RESEND_INTERVAL)
+        };
+
+        tokio::select! {
+            // Poll for outgoing requests, but only if we don't already _have_
+            // an outstanding request.
+            maybe_request = outgoing_request_rx.recv(), if outstanding_request.is_none() => {
+
+                // We only get `None` if the sender is closed, meaning the task
+                // holding that side exited. Nothing else will come down this
+                // channel, and things are likely borked. Bail out.
+                let request = match maybe_request {
+                    Some(r) => r,
+                    None => {
+                        debug!(log, "outgoing response channel closed, exiting");
+                        return;
+                    }
+                };
+                debug!(log, "received outgoing request: {request:?}");
+
+                // Store the outstanding request, sanity-checking that we really
+                // didn't have a prior one.
+                let old = outstanding_request.replace(request);
+                assert!(
+                    old.is_none(),
+                    "dequeued a new request while one is already outstanding!",
+                );
+
+                let request = outstanding_request.as_mut().unwrap();
+                match send_outgoing_request(&socket, &peer_addr, request, &mut send_buf).await {
+                    Ok(n) => debug!(log, "sent message"; "message" => ?request, "n_bytes" => n),
+                    Err(e) => error!(log, "failed to send message"; "reason" => ?e),
+                }
+            }
+
+            // If we _do_ have an outstanding request, we need to resend it
+            // periodically until we get a response. Wait for up to the resend
+            // interval of inactivity, and then possibly retry.
+            _ = retry_timeout, if outstanding_request.is_some() => {
+                debug!(log, "timed out without response, retrying");
+                let request = outstanding_request.as_mut().unwrap();
+                send_outgoing_request(&socket, &peer_addr, request, &mut send_buf).await.unwrap();
```

Comment:
Is this `.unwrap()` safe? If we previously failed to send up on line 367, we could end up in this branch and presumably fail to send again, right?

Reply:
No, it's not. Thanks for flagging. I'll use the same pattern as above, logging and continuing the loop.